Integrate final multi-agent project

SchwammDev opened the pull request:
This is the final project of the Udacity Reinforcement Learning course.

SchwammDev committed:
Add tennis environment to fetch script

This unity environment needs to be solved over the course of the 3rd
project.

SchwammDev committed:
Add multi agent handling to unity env adapter

To solve the third project the unity environment gym adapter needs to
handle multi agent environments properly, hence the adapter has been
extended. Also another test mode for the new environment has been added
to make sure it is handled properly. Additionally the make coverage
option has been extended to append reports for each environment run.
This way the code coverage report is more accurate. Currently I need to
run the unity environment tests in separate sessions because there is a
bug in the unity environment where they do not get cleaned up properly
so I need to fully restart the process. Finally I've extended the
exploration to allow for multiple episodes, so I could investigate the
Tennis environment properly.

SchwammDev committed:
Implement multi agent wrapper

This allows me to use the CLI interface to start and train a multi agent
setup. By wrapping existing agents, I make minimal changes to the code
and keep the architecture flexible.

SchwammDev committed:
Implement MADDPG agent

This way I can make more multi-agent specific changes to the DDPG
algorithm as I have for instance direct access to the critic model.
However, for now I just wrap observations and actions and treat it as a
single agent environment. This is easier than a multi agent approach and
will help me understand the specifics of the problem better. At the
moment the agent still has problems getting stuck in a local minima
where it often won't recover from. I suspect that better memory sampling
or improved exploration might help.

SchwammDev committed:
Improve exploration logic of MADDPG agent

By allowing preheat steps which are completely random the agent can
observe a more diverse set of states. Otherwise it wouldn't see that
much because it tends to crash early on into a local minimum. With these
trajectories in the replay buffer it seems to be better abel to get out
of it again. Additionally I've added static noise to the training. This
should also help in exploring and getting the agent out of local minima.
For now I just use a fixed value.

SchwammDev committed:
Remove MultiAgentWrapper

It is good practise to remove code that is not used anymore. This
particular implementation has been removed, because I decided to treat
multi agents as their own algorithms. Mostly because it is not so easy
to keep agents nicely separated. They often share replay buffers or
Q-Value models.

SchwammDev committed:
Implement keyboard interrupt and snapshots

I want to keep the state when of the agent when I interrupt training.
Often because it already reached a very high value. Also I wanted a way
of investigating agent behaviour during training to get a better feel
for why performance crashes early on. Additionally, to not loose well
trained agents I added a snapshot system.

SchwammDev committed:
Add final NDDPG agent solving the environment

A simple naive approach where I just used two separate DDPG agents for
each racket was able to solve the environment after sufficient training.
I compared it with a "single minded" agent that just trains and acts on
both rackets and they performance was equivalent. I suspect that the
naive approach is possible because. The agents agents do not influence
each other much. The action of one agent doesn't prevent or obscure
actions from the other, hence I think the each agent can treat the other
agent as part of the environment more easily. The problem of the crash
in performance early might stem from improperly initialized model
weights or a difficult reward function, where once at at the bottom
agents do not observe many useful actions anymore. Also performance
seems to be instable because watching the agents it seems like they
learn to play the ball at the exact same spot most of the time, creating
a careful balance. If the ball falls differently due to random
fluctuations the performance crashes again as agents mostly just observe
these "balanced" states. I suspect prioritized replay could improve
these issues.

SchwammDev committed:
Fix NDDPG tests and coverage report

With the new agent setup the basic learning tests failed, because it now
required different observation, action and reward shapes. Also the
multiple inheritance approach didn't work that well, because of the
order each gym wrapper was invoked. Also added a .coveragerc file to
make the coverage reports more accurate and to exclude main.py as it is
tested when using it directly.

SchwammDev committed:
Rename MADDPG to NDDPG

As the current implementation does not really follow the MADDPG
algorithm as described in the paper (link at the bottom), but rather
just trains two independent DDPG algorithms, I decided to rename it, to
avoid confusion.

https://arxiv.org/abs/1706.02275

SchwammDev committed:
Add first draft of the P3 README

This is needed for the submission of the final `multi-agent` project.

SchwammDev committed:
Add P3 report first draft

The report is necessary to finish the final project.

SwamyDev commented:
Don't forget to add the 3rd project report to the report file in the repository root

SchwammDev commented on doc/Report_p3_multiagent.md:
```diff
@@ -0,0 +1,80 @@
+The main "trick" was just to train both agents long enough to get them over the initial crashing phase. Once the algorithm reached peak performance around `>2`the algorithm would still fluctuate a lot, but not crash down to the minimum again. I'm suspecting this is due, the agents learning a careful balance where they would bounce the ball back end forth at the exact same position, for very long  ...
+
+## The DDPG Algorithm
+The algorithm is the same as described in [Report_p2_continuous](Report_p2_continuous.md). However, there are some minor tweaks to the exploration handling. I implement a similar setup as the `DDPG` algorithm described in [Spinning Up RL from OpenAI](https://spinningup.openai.com/en/latest/algorithms/ddpg.html). I now start with completely random "preheating" steps before letting the agent act, a ...
```
```suggestion
The algorithm is the same as described in [Report_p2_continuous.md](Report_p2_continuous.md). However, there are some minor tweaks to the exploration handling. I implement a similar setup as the `DDPG` algorithm described in [Spinning Up RL from OpenAI](https://spinningup.openai.com/en/latest/algorithms/ddpg.html). I now start with completely random "preheating" steps before letting the agent act, and then simply add gaussian noise with a fixed standard deviation to ensure continued exploration. 
```

SchwammDev committed:
Use max score of both multiple agents

Because I didn't read the instructions carefully enough, I've
accidentally took the mean score of both agents when computing the
overall score to determine if the environment was solved. However, the
maximum was required making the environment a little bit easier.

SchwammDev committed:
Add agent trained with max score

The previous agent has been compared to the mean accumulated score,
however, just the max was required and a new agent has been trained.

SchwammDev committed:
Update report and readme draft

Improved wording and corrected some spelling mistakes. Also corrected
for the score issue, where I've used the mean of both agent rewards
before. Now I use the correct max score.
